### DISCUSSION TITLE
posting a message to server from external process

### DISCUSSION THREAD

**Discussion by streamnsight** [miguelgrinberg/python-socketio#885] [Q&A] [answered]
Hello,
I use flask-socketio that uses python-socketio under the hood, to talk to clients.
I also have external python processes that send messages via the redis-adapter, and that works great.

As I debug this stuff, I can see the messages getting sent to the redis pubsub system, on the channel 'flask-socketio'
I see from the code that the data is pickled, and I see the raw pickled messages in redis pubsub. So far so good.

Now, I would like to also send messages from a Node JS service and I don't see an equivalent in node, so I am trying to send messages directly to the redis pubsub channel. however if I just copy a previous message and re-publish, the client(s) don't get it.

My question is: is there a way to directly send messages to redis pubsub to be picked up by the python subscribing process?
Has anyone done this before, or knows how that could work?

Thanks for help.

**Comment by miguelgrinberg** [comment]
You can also use JSON for the data, you do not need to implement the pickle encoding in your Node application. You can see an example of what the JSON object must contain here: https://github.com/miguelgrinberg/python-socketio/blob/main/src/socketio/pubsub_manager.py#L116-L120.

**Comment by streamnsight** [comment, marked as answer]
@miguelgrinberg 
Thanks for the quick reply, and thanks for the tip on using JSON.
I thought things had to be pickled because that's what the redis-manager does, but indeed after reviewing the listen code, I see it tries multiple methods, including json.loads.

I ended up with:

```
            const payload = {
                'method': 'emit', 
                'event': 'logs', 
                'data': {'type': 'fetching', 'state': 'failed'} , 
                'namespace': '/', 
                'room': null, 
                'skip_sid': null, 
                'callback': null
            };
            await redisClient.PUBLISH('flask-socketio', JSON.stringify(payload));
```

and that worked!
